topologies/binding/binding.go

```diff
@@ -299,3 +300,81 @@ func ports(tports []*opb.Port, bports []*bindpb.Port) (map[string]*binding.Port,
 	}
 	return portmap, nil
 }
+
+func (b *staticBind) reserveATE(ctx context.Context) error {
+	ates := b.resv.ATEs
+	for _, ate := range ates {
+		dialer, err := b.r.ixnetwork(ate.Name)
+		if err != nil {
+			return err
+		}
+		if _, err := b.ixSessionForATE(ctx, ate.Name, dialer); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (b *staticBind) releaseATE(ctx context.Context) error {
+	for ateName, ixw := range b.ateIxWeb {
+		dialer, err := b.r.ixnetwork(ateName)
+		if err != nil {
+			return err
+		}
+		ixs, ok := b.ateIxSession[ateName]
+		if ok && dialer.SessionId == 0 {
+			if err := ixw.IxNetwork().DeleteSession(ctx, ixs.ID()); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (b *staticBind) ixWebForATE(ctx context.Context, ateName string, d dialer) (*ixweb.IxWeb, error) {
+	b.muIxWeb.Lock()
```

Comment:
Is the question here: can a binding's implementation of Reserve and Release be called concurrently? No they cannot. If we find a case in which they are, that would be a bug on Ondatra to prevent the binding from being called in that way. Our philosophy is that the framework ought to relieve the burden on the binding implementation as much as possible.

Reply:
Okay, @mingyangcisco please remove the mutex.

I also decided to take a moment and write a blog post about it with example code, in case it will be helpful in the future: https://lifecs.likai.org/2022/04/the-correct-way-to-lock-get-or-create.html

Reply:
Cool. Thanks @liulk. I just removed the locks and will go through your blog.